src/app/controllers/tour.controller.spec.ts

```diff
@@ -0,0 +1,176 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { UserDto } from '../../core/dtos/user';
+import { getRepositoryToken } from '@nestjs/typeorm';
+import { TourController } from './tour.controller';
+import { TourService } from '../../core/services/tour.service';
+import { Tour } from '../../infrastructure/entities/tour.entity';
```

Comment:
As already discussed: This violates the architecture as the app layer should never communicate directly with the infrastructure layer.

Reply:
@Saela Yes, I thought about that after our discussion. However - does it really communicate with the infrastructure layer? I mean we only import the mocking components and inject it, so the actual communication with the infrastructure layer still happens via the service below.

(Of course, we also see the issue with our controller tests - we're basically just wrapping the tests of our service in another layer which does not really add anything new, apart from testing whether the controller calls the right service method 🤷 )

Reply:
@Tugark Well no, I don't think you can call it communication in the sense that they exchange data. But in a layered structure if you need to share classes or utility stuff between layers you'd normally add them in a shared module or the core layer. Otherwise you break the concept of layering. In e.g. a .net application if you'd wanted to use some entities from the infrastructure layer in your app layer the IDE would tell you to import the module so it makes it more clear to you that this should be refactored. With nestjs this is a bit different and makes it harder to keep the layers clean.

I argue for a revisit of the architecture. After working with nest some time I think using separate modules for the different concerns is a better approach.

And yes as for the controller tests: I would use the nice e2e functionality nestjs provides you. This makes more sense and will be a more useful solution than unit testing controllers.

Reply:
@Saela I get what you're saying and I am all for revisiting the architecture as you said. I also believe that we should use nest's module isolation which encapsulates the layered architecture for each module (and it also allows us to properly use nest's generation capabilities, which are rather painful to use with our current architecture).

However, I think with regards to the layering issue, that would still persists. Imagine we have a module `tours` that has these files in it. For the tests, we'd still have to import the mock repository, which then would still be in violation of layering principles, although within its own module. Yet in order for our test to work, we have to mock the repository and inject it for our service to work which is called by the tests. Our module itself does **not** violate the principles of layering itself, since we're not importing the infrastructure components anywhere within our Controllers - that's just for the setup of the tests. Or am I not getting what you're hinting at? :D

Besides, there would be 2 solutions for this issue I guess:

* What you suggested with a dedicated testing module that does all this setup work so it does not have to be made within the controller test
* Come to think of it, I believe my ControllerTest is mocking the wrong part - it should only mock the underlying service and not the repository. Could that be the whole issue? :D